public/app/core/components/NestedFolderPicker/api.ts

```diff
@@ -0,0 +1,206 @@
+  //   b) use them to select the responses out of the state
+  const requestsRef = useRef<ListFoldersRequest[]>([]);
+
+  const state = useSelector((rootState: RootState) => {
+    return listAllFoldersSelector(rootState, requestsRef.current);
+  });
+
+  // Loads the next page of folders for the given parent UID by inspecting the
+  // state to determine what the next page is
+  const requestNextPage = useCallback(
+    (parentUid: string | undefined) => {
+      const pages = parentUid ? state.pagesByParent[parentUid] : state.rootPages;
+      const [fullyLoaded, pageNumber] = getPagesLoadStatus(pages ?? []);
+      if (fullyLoaded) {
+        return;
+      }
+
+      const args = { parentUid, page: (pageNumber ?? 0) + 1, limit: PAGE_SIZE };
+      const promise = dispatch(browseDashboardsAPI.endpoints.listFolders.initiate(args));
+
+      // It's important that we create a new array so we can correctly memoize with it
+      requestsRef.current = requestsRef.current.concat([promise]);
+    },
+    [state, dispatch]
+  );
+
+  // Unsubscribe from all requests when the component is unmounted
+  useEffect(() => {
+    return () => {
+      for (const req of requestsRef.current) {
+        req.unsubscribe();
+      }
+    };
+  }, []);
+
+  // Convert the individual responses into a flat list of folders, with level indicating
+  // the depth in the hierarchy.
+  const treeList = useMemo(() => {
+    if (!isBrowsing) {
+      return [];
+    }
+
+    function createFlatList(
+      parentUid: string | undefined,
+      pages: ListFoldersQuery[],
+      level: number
+    ): Array<DashboardsTreeItem<DashboardViewItemWithUIItems>> {
+      const flatList = pages.flatMap((page) => {
+        const pageItems = page.data ?? [];
+
+        return pageItems.flatMap((item) => {
+          const folderIsOpen = openFolders[item.uid];
+
+          const flatItem: DashboardsTreeItem<DashboardViewItemWithUIItems> = {
+            isOpen: Boolean(folderIsOpen),
+            level: level,
+            item: {
+              kind: 'folder' as const,
+              title: item.title,
+              uid: item.uid,
+            },
+          };
+
+          const childPages = folderIsOpen && state.pagesByParent[item.uid];
+          if (childPages) {
+            const childFlatItems = createFlatList(item.uid, childPages, level + 1);
+            return [flatItem, ...childFlatItems];
+          }
+
+          return flatItem;
+        });
+      });
+
+      const [fullyLoaded] = getPagesLoadStatus(pages);
+      if (!fullyLoaded) {
+        flatList.push(...getPaginationPlaceholders(PAGE_SIZE, parentUid, level));
+      }
+
+      return flatList;
+    }
+
+    const rootFlatTree = createFlatList(undefined, state.rootPages, 1);
+    rootFlatTree.unshift(ROOT_FOLDER_ITEM);
+
+    return rootFlatTree;
+  }, [state, isBrowsing, openFolders]);
+
+  return [treeList, state.isLoading, requestNextPage] as const;
```

Comment:
I think using object return would be better here so you don't need to remember the order.